core/src/main/java/org/infinispan/commands/irac/IracCleanupTombstoneCommand.java

```diff
@@ -0,0 +1,137 @@
+package org.infinispan.commands.irac;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+import org.infinispan.commons.util.Util;
+import org.infinispan.distribution.DistributionInfo;
+import org.infinispan.factories.ComponentRegistry;
+import org.infinispan.metadata.impl.IracMetadata;
+import org.infinispan.remoting.responses.ValidResponse;
+import org.infinispan.remoting.rpc.RpcManager;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.remoting.transport.ValidSingleResponseCollector;
+import org.infinispan.util.ByteString;
+import org.infinispan.util.concurrent.CompletableFutures;
+import org.infinispan.xsite.BackupReceiver;
+import org.infinispan.xsite.XSiteReplicateCommand;
+
+public class IracCleanupTombstoneCommand extends XSiteReplicateCommand<Boolean> {
+
+   public static final byte COMMAND_ID = 37;
+
+   private Object key;
+   private IracMetadata tombstone;
+
+   @SuppressWarnings("unused")
+   public IracCleanupTombstoneCommand() {
+      super(COMMAND_ID, null);
+   }
+
+   public IracCleanupTombstoneCommand(ByteString cacheName) {
+      super(COMMAND_ID, cacheName);
+   }
+
+   public IracCleanupTombstoneCommand(ByteString cacheName, Object key, IracMetadata tombstone) {
+      super(COMMAND_ID, cacheName);
+      this.key = key;
+      this.tombstone = tombstone;
+   }
+
+   @Override
+   public ByteString getCacheName() {
+      return cacheName;
+   }
+
+   @Override
+   public CompletableFuture<Boolean> invokeAsync(ComponentRegistry registry) {
+      if (tombstone == null) {
+         return registry.getIracManager().running().containsKey(key) ?
+               CompletableFutures.completedTrue() :
+               CompletableFutures.completedFalse();
+      }
+      registry.getIracTombstoneCleaner().running().removeTombstone(key, tombstone);
+      return CompletableFutures.completedNull();
+   }
+
+   @Override
+   public byte getCommandId() {
+      return COMMAND_ID;
+   }
+
+   @Override
+   public CompletionStage<Boolean> performInLocalSite(ComponentRegistry registry, boolean preserveOrder) {
+      DistributionInfo distribution = registry.getDistributionManager().getCacheTopology().getDistribution(key);
+      if (distribution.isPrimary()) {
+         return registry.getIracManager().running().containsKey(key) ?
+               CompletableFutures.completedTrue() :
+               CompletableFutures.completedFalse();
+      } else {
+         RpcManager manager = registry.getRpcManager().running();
+         return manager.invokeCommand(distribution.primary(), this, new BooleanResponseCollector(), manager.getSyncRpcOptions());
```

Comment:
`performInLocalSite` is invoked in the site master. If the site master is the primary owner, it executes the command; otherwise, it forwards to the primary owner.
In any case, a boolean return value is expected.

Reply:
Ah sorry, I confused the methods and thought the `else` block was for `tombstone != null`